lib/client/src/jup.rs

```diff
@@ -0,0 +1,325 @@
+            (
+                order_amounts.total_quote_taken_native,
+                order_amounts.total_base_taken_native,
+            )
+        } else {
+            (
+                order_amounts.total_base_taken_native,
+                order_amounts.total_quote_taken_native,
+            )
+        };
+
+        Ok(Quote {
+            in_amount,
+            out_amount,
+            fee_mint: self.market.quote_mint,
+            fee_amount: order_amounts.fee,
+            price_impact_pct: order_amounts.price_impact.into(),
+            ..Quote::default()
+        })
+    }
+}
+
+pub struct Amounts {
+    pub total_base_taken_native: u64,
+    pub total_quote_taken_native: u64,
+    pub fee: u64,
+    pub price_impact: i64,
+}
+
+fn iterate_book(
+    book: Orderbook,
+    order: &Order,
+    market: &Market,
+    oracle_price: I80F48,
+    now_ts: u64,
+) -> Result<Amounts> {
+    let side = order.side;
+    let mut limit = MAXIUM_TAKEN_ORDERS;
+
+    let other_side = side.invert_side();
+    let oracle_price_lots = market.native_price_to_lot(oracle_price)?;
+    let (price_lots, _) = order.price(now_ts, oracle_price_lots, &book)?;
+
+    let order_max_quote_lots = match side {
+        Side::Bid => market.subtract_taker_fees(order.max_quote_lots_including_fees),
+        Side::Ask => order.max_quote_lots_including_fees,
+    };
+
+    let mut remaining_base_lots = order.max_base_lots;
+    let mut remaining_quote_lots = order_max_quote_lots;
+
+    let mut first_price = 0_i64;
+    let mut last_price = 0_i64;
+
+    let opposing_bookside = book.bookside(other_side);
+    for (index, best_opposing) in opposing_bookside
+        .iter_valid(now_ts, oracle_price_lots)
+        .enumerate()
+    {
+        if remaining_base_lots == 0 || remaining_quote_lots == 0 {
+            break;
+        }
+
+        if !best_opposing.is_valid() {
+            continue;
+        }
+        if index == 0 {
+            first_price = best_opposing.price_lots;
+        } else {
+            last_price = best_opposing.price_lots;
+        }
+
+        let best_opposing_price = best_opposing.price_lots;
+
+        if !side.is_price_within_limit(best_opposing_price, price_lots) || limit == 0 {
+            break;
+        }
+
+        let max_match_by_quote = remaining_quote_lots / best_opposing_price;
+        if max_match_by_quote == 0 {
+            break;
+        }
+
+        let match_base_lots = remaining_base_lots
+            .min(best_opposing.node.quantity)
+            .min(max_match_by_quote);
+        let match_quote_lots = match_base_lots * best_opposing_price;
+
+        remaining_base_lots -= match_base_lots;
+        remaining_quote_lots -= match_quote_lots;
+        assert!(remaining_quote_lots >= 0);
+
+        limit -= 1;
+    }
+
+    let total_quote_lots_taken = order_max_quote_lots - remaining_quote_lots;
+    let total_base_lots_taken = order.max_base_lots - remaining_base_lots;
+
+    assert!(total_quote_lots_taken >= 0);
```

Comment:
not sure if you want to panic here instead of setting `quote.not_enough_liquidity = true`